aea/cli/ipfs_hash.py

```diff
@@ -0,0 +1,491 @@
+
+    :param package_type: the package type.
+    :param package_path: the path to the package root.
+    :return: the configuration object.
+    """
+    configuration_class = type_to_class_config[package_type]
+    configuration_filepath = (
+        package_path / configuration_class.default_configuration_filename
+    )
+
+    loader = ConfigLoaders.from_package_type(package_type)
+    with configuration_filepath.open() as fp:
+        configuration_obj = loader.load(fp)
+    configuration_obj._directory = package_path  # pylint: disable=protected-access
+    return cast(PackageConfiguration, configuration_obj)
+
+
+def assert_hash_consistency(fingerprint: Dict[str, str], path_prefix: Path) -> None:
+    """
+    Check that our implementation of IPFS hashing for a package is correct against the true IPFS.
+
+    :param fingerprint: the fingerprint dictionary.
+    :param path_prefix: the path prefix to prepend.
+    """
+    # confirm ipfs only generates same hash:
+    for file_name, ipfs_hash in fingerprint.items():
+        path = path_prefix / file_name
+        expected_ipfs_hash = HASH_TOOL.hash_file(str(path), wrap=False)
+
+        if expected_ipfs_hash != ipfs_hash:
+            raise ValueError("WARNING, hashes don't match for: {}".format(path))
+
+
+def _replace_fingerprint_non_invasive(
+    fingerprint_dict: Dict[str, str], text: str
+) -> str:
+    """
+    Replace the fingerprint in a configuration file (not invasive).
+
+    We need this function because libraries like `yaml` may modify the
+    content of the .yaml file when loading/dumping. Instead,
+    working with the content of the file gives us finer granularity.
+
+    :param fingerprint_dict: the fingerprint dictionary.
+    :param text: the content of a configuration file.
+    :return: the updated content of the configuration file.
+    """
+
+    def to_row(x: Tuple[str, str]) -> str:
+        return x[0] + ": " + x[1]
+
+    replacement = "\nfingerprint:\n  {}\n".format(
+        "\n  ".join(map(to_row, sorted(fingerprint_dict.items())))
+    )
+    return re.sub(r"\nfingerprint:\W*\n(?:\W+.*\n)*", replacement, text)
+
+
+def compute_fingerprint(  # pylint: disable=unsubscriptable-object
+    package_path: Path, fingerprint_ignore_patterns: Optional[Collection[str]],
+) -> Dict[str, str]:
+    """
+    Compute the fingerprint of a package.
+
+    :param package_path: path to the package.
+    :param fingerprint_ignore_patterns: filename patterns whose matches will be ignored.
+    :return: the fingerprint
+    """
+    fingerprint = _compute_fingerprint(
+        package_path, ignore_patterns=fingerprint_ignore_patterns,
+    )
+    assert_hash_consistency(fingerprint, package_path)
+    return fingerprint
+
+
+def update_fingerprint(configuration: PackageConfiguration) -> None:
+    """
+    Update the fingerprint of a package.
+
+    :param configuration: the configuration object.
+    :return: None
+    """
+    # we don't process agent configurations
```

Comment:
I can see this from the code. Would be more useful to know why we don't; please explain

Reply:
Agent package only contains only 2 files, a README file and the agent-config.yaml. Both of which doesn't need to be fingerprinted.

Reply:
That's actually wrong. We do need to fingerprint them. We fingerprint `README.md` in other packages too and then an agent package can have more files, e.g. if you add a custom decision maker. So we should not skip that!

Reply:
Then open one please...